lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerCodeTransform\CodeAction;
+
+use Amp\Promise;
+use Amp\Success;
+use Microsoft\PhpParser\Node\Expression\CallExpression;
+use Microsoft\PhpParser\Node\Expression\MemberAccessExpression;
+use Microsoft\PhpParser\Node\Expression\ScopedPropertyAccessExpression;
+use Microsoft\PhpParser\Parser;
+use Microsoft\PhpParser\Token;
+use Phpactor\Extension\LanguageServerBridge\Converter\PositionConverter;
+use Phpactor\Extension\LanguageServerCodeTransform\LspCommand\GenerateMethodCommand;
+use Phpactor\LanguageServerProtocol\CodeAction;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\Diagnostic;
+use Phpactor\LanguageServerProtocol\DiagnosticSeverity;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServer\Core\CodeAction\CodeActionProvider;
+use Phpactor\LanguageServer\Core\Diagnostics\DiagnosticsProvider;
+use Phpactor\TextDocument\ByteOffset;
+use function Amp\call;
+use function array_filter;
+
+class GenerateMethodProvider implements DiagnosticsProvider, CodeActionProvider
+{
+    public const KIND = 'quickfix.generate_method';
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(Parser $parser)
+    {
+        $this->parser = $parser;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function kinds(): array
+    {
+        return [
+             self::KIND
+         ];
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function provideDiagnostics(TextDocumentItem $textDocument): Promise
+    {
+        return new Success($this->getDiagnostics($textDocument));
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range): Promise
+    {
+        return call(function () use ($textDocument, $range) {
+            if ($range->start != $range->end) {
+                return [];
+            }
+
+            $diagnostics = $this->getDiagnostics($textDocument);
+
+            if (empty($diagnostics)) {
+                return [];
+            }
+
+            $diagnostics = array_values(
+                array_filter($diagnostics, function (Diagnostic $diag) use ($range) {
+                    return (
+                        $diag->range->start->line <= $range->start->line &&
+                        $diag->range->start->character <= $range->start->character &&
+                        $diag->range->end->line >= $range->end->line &&
+                        $diag->range->end->character >= $range->end->character
+                    );
+                })
+            );
+
```

Comment:
The above probably needs to change to a code similar to what diagnostics does. I am generally fearful of performance similar to that of Import Names where it gets really slow. I wonder if we can cache the result of the diagnostics and use that to see whether we're on one? If the cache is not there we can always get them again.